composer/utils/object_store/mlflow_object_store.py

```diff
@@ -0,0 +1,392 @@
+# Copyright 2022 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""MLFlow Artifacts object store."""
+
+from __future__ import annotations
+
+import logging
+import os
+import pathlib
+import uuid
+from typing import Callable, List, Optional, Tuple, Union
+
+from composer.utils import dist
+from composer.utils.import_helpers import MissingConditionalImportError
+from composer.utils.object_store.object_store import ObjectStore, ObjectStoreTransientError
+
+__all__ = ['MLFlowObjectStore']
+
+MLFLOW_DATABRICKS_TRACKING_URI = 'databricks'
+MLFLOW_DBFS_PATH_PREFIX = 'databricks/mlflow-tracking/'
+
+DEFAULT_MLFLOW_EXPERIMENT_NAME = 'mlflow-object-store'
+
+PLACEHOLDER_EXPERIMENT_ID = 'MLFLOW_EXPERIMENT_ID'
+PLACEHOLDER_RUN_ID = 'MLFLOW_RUN_ID'
+
+log = logging.getLogger(__name__)
+
+
+def _wrap_mlflow_exceptions(uri: str, e: Exception):
+    """Wraps retryable MLFlow errors in ObjectStoreTransientError for automatic retry handling."""
+    from mlflow.exceptions import (ABORTED, DATA_LOSS, DEADLINE_EXCEEDED, ENDPOINT_NOT_FOUND, INTERNAL_ERROR,
+                                   INVALID_STATE, NOT_FOUND, REQUEST_LIMIT_EXCEEDED, RESOURCE_DOES_NOT_EXIST,
+                                   RESOURCE_EXHAUSTED, TEMPORARILY_UNAVAILABLE, ErrorCode, MlflowException)
+
+    # https://github.com/mlflow/mlflow/blob/master/mlflow/exceptions.py for used error codes.
+    # https://github.com/mlflow/mlflow/blob/master/mlflow/protos/databricks.proto for code descriptions.
+    retryable_server_codes = [
+        ErrorCode.Name(code) for code in [
+            DATA_LOSS,
+            INTERNAL_ERROR,
+            INVALID_STATE,
+            TEMPORARILY_UNAVAILABLE,
+            DEADLINE_EXCEEDED,
+        ]
+    ]
+    retryable_client_codes = [ErrorCode.Name(code) for code in [ABORTED, REQUEST_LIMIT_EXCEEDED, RESOURCE_EXHAUSTED]]
+    not_found_codes = [ErrorCode.Name(code) for code in [RESOURCE_DOES_NOT_EXIST, NOT_FOUND, ENDPOINT_NOT_FOUND]]
+
+    if isinstance(e, MlflowException):
+        error_code = e.error_code  # pyright: ignore
```

Comment:
I think this might actually be a pyright bug since it should be able to infer the type correctly from `if isinstance(e, MlflowException)` (VSCode/pylance infers correctly) but instead I get `error: Cannot access member "error_code" for type "Exception"`

I tried both `assert isinstance(e, MlflowException)` and `assert type(e) is MlflowException`, but pyright won't recognize it 😢

Reply:
what about `assert hasattr(e, 'error_code')`? if that doesn't work, oh well, thanks for trying

Reply:
no luck...will keep the ignore I guess